### ISSUE TITLE
Should not report an error on "null" when using in type position

### ISSUE THREAD

**Issue by yuit** [palantir/tslint#1265]
We are trying to add linting to our library source files and run into issue below.

### Bug Report (delete this for feature requests)

- __TSLint version__: `@next`
- __TypeScript version__: `latest source-code built in master branch`
- __Running TSLint via__: Node.js API 

#### TypeScript code being linted

```ts
match(regexp: string): RegExpMatchArray | null;
```

with `tslint.json`: 

```json
     ....
     "no-null-keyword": true,
```

#### Actual behavior
Get an error `Use 'undefined' instead of 'null'`

#### Expected behavior

There shouldn't be an error when using `null` in type position

**Comment by JKillian** [comment]
Good catch, `null` should be allowed as a type

**Comment by adidahiya** [comment]
specifically, this has to do with the `no-null-keyword` rule

**Comment by JKillian** [comment]
Two quick questions @yuit:

* Do you know an easy way when dealing with the AST to tell if a node is being used as a type?
* Does TS produce the same AST no matter if the `--strictNullChecks` flag is on or not?

**Comment by JKillian** [comment]
Okay, as far as I understand [from this](https://github.com/Microsoft/TypeScript/pull/7140#issue-134699594), no matter what the compiler options, `null` is still allowed in a type position. So, TSLint should be able to disregard whether `--strictNullChecks` is being used by the end user which is very good.

So question two from above taken care of I think 😃

**Comment by JKillian** [comment]
As for my first question, although a little different than the top-down approach TSLint normally uses, seemed easiest to just traverse through the ancestors of each found null keyword and check if any ancestor has a type kind. Theoretically this could be re-traversing a lot of nodes, but in this context the issue shouldn't arise.

**Comment by yuit** [comment]
@JKillian I agree to check that `null` is used in type position, you will have to walk up the parents to check if any has a type kind. However, as you mentions it shouldn't be too much performance hit since the level of traversing shouldn't be too deep

**Comment by JKillian** [comment]
Great, fix is merged! This will be in a release shortly

**Comment by yuit** [comment]
Thanks ! :+1:

**Comment by JKillian** [comment]
Released in `v3.10.0-dev.3` which is what `tslint@next` currently points to, cheers!
